### PR TITLE
Feat(forms):a11y updates

### LINKED ISSUE #587: dive deeper into valid/invalid inputs for forms
This is related to #504 

Based on @jgiardino's review I think we need to break out some of the accessibility work to test things further. 

This will require further discussion but it revolves around more extensive testing and what that looks like in the workspace.

### PR DESCRIPTION
**Pull request by matthewcarleton**
Closes #587 

These are changes that were required after doing further testing with voiceover. It also includes changes that @mattnolting requested in #587. I've kept the nested conditional, although looking at it now I'm not sure what it is doing exactly.

Things to note:

- After moving to styling `aria-invalid="true"` instead of `:invalid` I'm wondering if we need the `pf-m-invalid` at all.
- I've added the modifier `.pf-m-inactive` to handle inactive helper text. 
- I've added notes for `aria-errormessage` and `aria-invalid=true` in the form accessibility docs. Not sure if this is how we want to document the requirements of Javascript.

### CODE REVIEW

File: src/patternfly/components/Form/examples/form-text-input-example.hbs [hunk 1]
```diff
@@ -1,39 +1,39 @@
...
   {{#> form-group}}
     {{#> form-label for="textInput2"}}Input with placeholder{{/form-label}}
     {{#> form-controls controlType="input" input="true" type="text" name="textInput2" id="textInput2" placeholder="Placeholder input example"}}{{/form-controls}}
   {{/form-group}}
     {{#> form-group}}
-    {{#> form-label for="textInput2"}}Readonly Input{{/form-label}}
-    {{#> form-controls controlType="input" input="true" type="text" name="textInput2" id="textInput2" readonly="true" value="Readonly input example"}}{{/form-controls}}
+    {{#> form-label for="textInput3"}}Readonly Input{{/form-label}}
+    {{#> form-controls controlType="input" input="true" type="text" name="textInput3" id="textInput3" readonly="true" value="Readonly input example"}}{{/form-controls}}
   {{/form-group}}
```
Comment (mattnolting) [inline src/patternfly/components/Form/examples/form-text-input-example.hbs]: The readonly input currently allows text input. Need to add `readonly` to input

File: src/patternfly/components/Form/docs/code.md [hunk 2]
```diff
@@ -7,7 +7,7 @@
 | `required aria-required="true"` | `<input>`, `<select>`, `<textarea>` | Required fields must include these attributes. |
 | `id="{helper_text_id}"` | `.pf-c-form__helper-text` | Form fields with related `.pf-c-form__helper-text` require this attribute. Usage `<p class="pf-c-form__helper-text" id="{helper_text_id}" ...>`.  |
 | `aria-describedby="{helper_text_id}"` | `<input>`, `<select>`, `<textarea>` | Form fields with related `.pf-c-form__helper-text` require this attribute. Usage `<input aria-describedby="{helper_text_id}" ...>`.  |
```
Comment (mattnolting) [inline src/patternfly/components/Form/docs/code.md]: Is `...` in `<input aria-describedby="{helper_text_id}" ...>` intentional? I assume it denotes 'other' attributes, but idk if it's necessary.

File: src/patternfly/components/FormControls/form-controls.hbs [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+{{#if placeholder}} placeholder="{{placeholder}}"{{/if}}
+{{#if value}} value="{{value}}"{{/if}}
+{{#if name}} name="{{name}}"{{/if}}
+{{#if aria-describedby}} aria-describedby="{{aria-describedby}}"{{/if}}
+{{#if aria-label}} aria-label="{{aria-label}}" {{/if}}
+{{#if aria-errormessage}} aria-errormessage="{{aria-errormessage}}"{{/if}}
+{{#if required}}required aria-required="true"{{#if aria-invalid}}aria-invalid="true"{{/if}}{{/if}}
+{{#if disabled}} disabled{{/if}}
+{{#if id}}id="{{id}}"{{/if}}
+{{#if multiple}} multiple="true"{{/if}}
```
Comment (mattnolting) [inline src/patternfly/components/FormControls/form-controls.hbs]: {{#if readonly}} readonly{{/if}} (mentioned above)

File: src/patternfly/components/InputGroup/input-group-text.hbs [hunk 4]
```diff
@@ -2,6 +2,9 @@
 {{#if aria-label}}
     aria-label="{{aria-label}}"
 {{/if}}
+{{#if id}}
+    id="{{id}}"
+{{/if}}
 >
```
Comment (mattnolting) [inline src/patternfly/components/InputGroup/input-group-text.hbs]: I think you could handle the `id` with `input-group-text--attribute`

File: src/patternfly/components/Button/button.hbs [hunk 5]
```diff
@@ -1,6 +1,9 @@
 <button class="pf-c-button{{#if btnClass}} {{btnClass}}{{/if}}{{#if button--disabled}} pf-m-disabled{{/if}}"
   {{#if btnAttributes}}
     {{{ btnAttributes }}} 
+  {{/if}}
+  {{#if id}}
+    id="{{id}}" 
   {{/if}}
```
Comment (mattnolting) [inline src/patternfly/components/Button/button.hbs]: I think you could handle the `id` with `button--attribute`

File: src/patternfly/components/Form/examples/form-text-input-example.hbs [hunk 6]
```diff
@@ -1,39 +1,39 @@
...
-  {{/form-group}} --}}
 
   {{#> form-group}}
-    {{#> form-label for="textInput4"}}Error state input with helper text{{/form-label}}
-    {{#> form-controls controlType="input" input="true" pf-c-form-control--modifier="pf-m-invalid" reqiured="true" aria-invalid="true" aria-errormessage="helper3" type="text" name="textInput4" id="textInput4"  aria-label="Error state text input" aria-describedby="helper3"}}{{/form-controls}}
-    {{#> form-helper-text pf-c-form__helper--modifier="pf-m-error" id="helper3"}}This is helper text{{/form-helper-text}}
+    {{#> form-label for="textInput7"}}Inactive Error state input with inactive helper text{{/form-label}}
+    {{#> form-controls controlType="input" input="true" type="text" name="textInput7" id="textInput7" }}{{/form-controls}}
+    {{#> form-helper-text pf-c-form__helper--modifier="pf-m-error pf-m-inactive" id="helper2" aria-live="polite"}}This is inactive helper text{{/form-helper-text}}
+  {{/form-group}}
```
Comment (mattnolting) [inline src/patternfly/components/Form/examples/form-text-input-example.hbs]: Exciting to see the introduction of `aria-live`! 👍

File: src/patternfly/components/InputGroup/styles.scss [hunk 7]
```diff
@@ -29,7 +29,8 @@
     margin-left: -1px;
   }
   /* stylelint-enable */
-  .pf-m-invalid {
+
+  .pf-c-form-control[aria-invalid="true"] {
     margin-left: 0;
```
Comment (michael-coker) [inline src/patternfly/components/InputGroup/styles.scss]: I think it looks better without this and the red border is on top of the grey border :)

File: src/patternfly/components/InputGroup/styles.scss [hunk 8]
```diff
@@ -29,7 +29,8 @@
...
 
     &:not(:last-child) {
```
Comment (michael-coker) [inline src/patternfly/components/InputGroup/styles.scss]: same here, I think it looks better without the margin and with `position: relative` (to give it a z-index) so the red border is on top on the right side.

File: src/patternfly/components/InputGroup/examples/input-group-example.hbs [hunk 9]
```diff
@@ -1,45 +1,45 @@
...
+  {{#> form-controls controlType="textarea" name="textarea2" id="textarea2" aria-label="textarea with button" aria-describedby="textAreaButton2" }}
 {{/form-controls}}
-    {{#> button btnClass="pf-m-tertiary"}} Button {{/button}}
+    {{#> button btnClass="pf-m-tertiary" id="textAreaButton2"}} Button {{/button}}
 {{/input-group}}
 <br>
 <br>
 {{#> input-group }}
-  {{#> dropdown id="dropdown-example-collapsed"}}
+  {{#> dropdown }}
```
Comment (michael-coker) [inline src/patternfly/components/InputGroup/examples/input-group-example.hbs]: Why remove the `id`? The dropdown component assumes that attribute exists, so you have empty `id=""` on the button and `aria-labelledby=""` on the dropdown menu.

File: src/patternfly/components/InputGroup/examples/input-group-example.hbs [hunk 10]
```diff
@@ -1,45 +1,45 @@
...
 <br>
 <br>
 {{#> input-group }}
-  {{#> form-controls controlType="input" input="true" type="text" id="textInput4" name="textInput4" }}
+  {{#> form-controls controlType="input" input="true" type="text" id="textInput4" name="textInput4" aria-label="another input with dropdown and button" aria-describedby="inputDropdownButton2" }}
   {{/form-controls}} 
-  {{#> button btnClass="pf-m-secondary"}} Button {{/button}}
-  {{#> dropdown id="dropdown-example-collapsed"}}
+  {{#> button btnClass="pf-m-secondary" id="inputDropdownButton2"}} Button {{/button}}
+  {{#> dropdown }}
```
Comment (michael-coker) [inline src/patternfly/components/InputGroup/examples/input-group-example.hbs]: same as above - creates an empty `id` and `aria-labelledby`

File: src/patternfly/components/Form/examples/form-text-input-example.hbs (hunk 1 above)
Comment (matthewcarleton) [reply src/patternfly/components/Form/examples/form-text-input-example.hbs]: doh

File: src/patternfly/components/InputGroup/examples/input-group-example.hbs (hunk 9 above)
Comment (matthewcarleton) [reply src/patternfly/components/InputGroup/examples/input-group-example.hbs]: ah, I was just getting rid of the duplicate ID's I'll add them back in.

File: src/patternfly/components/InputGroup/examples/input-group-example.hbs (hunk 10 above)
Comment (matthewcarleton) [reply src/patternfly/components/InputGroup/examples/input-group-example.hbs]: 👍

File: src/patternfly/components/FormControls/form-controls.hbs (hunk 3 above)
Comment (mattnolting) [reply src/patternfly/components/FormControls/form-controls.hbs]: mhmm, you had the attribute in the example. prod just got dropped.

**Comment by michael-coker** [comment]
When you rebase, you should pull in the change that moves `components/FormControls` to `components/FormControl`, if you want to go ahead and make that change. That's needed to fix the form control links under the codeblocks where form controls are in the list of "components used:".

And I missed this in that change, but we should also move `form-controls.hbs` to `form-control.hbs`, change `{{#> form-controls` to `{{#> form-control`, and update the examples to be consistent.

File: src/patternfly/components/Form/docs/code.md (hunk 2 above)
Comment (matthewcarleton) [reply src/patternfly/components/Form/docs/code.md]: good question, that was from the work that was already done on forms ;) Happy to remove.

File: src/patternfly/components/FormControl/examples/form-control-select-example.hbs [hunk 11]
```diff
@@ -1,13 +1,13 @@
-{{#> form-controls controlType="select" id="select1" name="selectexample" aria-label="select example"}}
+{{#> form-control controlType="select" id="selectExample1" name="selectExample" aria-label="select example"}}
```
Comment (michael-coker) [inline src/patternfly/components/FormControl/examples/form-control-select-example.hbs]: the `name` should probably match the `id` to be consistent.

File: src/patternfly/components/FormControl/examples/form-control-select-example.hbs [hunk 12]
```diff
@@ -16,8 +16,8 @@
     <option value="Option 3">The third option</option>
     <option value="Option 4">The fourth option</option>
   </optgroup>
-{{/form-controls}}
-{{#> form-controls controlType="select" pf-c-form-control--modifier="pf-m-invalid" required="true" aria-errormessage="select2ErrorText" id="select2" name="selectexample2" aria-label="select group example"}}
+{{/form-control}}
+{{#> form-control controlType="select" required="true" aria-invalid="true" id="selectExample3" name="selectExample3" aria-label="select group example"}}
```
Comment (michael-coker) [inline src/patternfly/components/FormControl/examples/form-control-select-example.hbs]: the `aria-label` should indicate this is an error state to be consistent with the other examples.

**Pull request merged**